saga/sagaState_prop_test.go

```diff
@@ -0,0 +1,215 @@
+	properties.Property("Empty String is Invalid TaskId", prop.ForAll(
+		func(id string) bool {
+			err := validateTaskId(id)
+
+			return (err != nil && id == "") || err == nil
+		},
+		gen.AnyString(),
+	))
+
+	properties.TestingRun(t)
+}
+
+func Test_ValidateUpdateSagaState(t *testing.T) {
+	parameters := gopter.DefaultTestParameters()
+	parameters.MinSuccessfulTests = 1000
+	properties := gopter.NewProperties(parameters)
+
+	properties.Property("StartSaga message is never valid on an already started saga", prop.ForAll(
+		func(state *SagaState, data []byte) bool {
+			msg := MakeStartSagaMessage(state.SagaId(), data)
+			newState, err := updateSagaState(state, msg)
+
+			return err != nil && newState == nil
+		},
+		GenSagaState(),
+		gen.SliceOf(gen.UInt8()),
+	))
+
+	// EndSaga messages are valid if a saga has not been Aborted and all StartTask have EndTask messages
+	// If a saga has been aborted all StartTask messages must have corresponding StartCompTask / EndCompTask messages
+	// for an EndSaga message to be valid.
+	properties.Property("EndSaga message is Valid or returns an error", prop.ForAll(
+		func(state *SagaState) bool {
+
+			msg := MakeEndSagaMessage(state.SagaId())
+			newState, err := updateSagaState(state, msg)
+
+			validTransition := true
+			for _, id := range state.GetTaskIds() {
+
+				// if aborted all comp tasks must be completed for all tasks started
+				// if not aborted all started tasks must be completed
+				if state.IsSagaAborted() {
+					if !(state.IsTaskStarted(id) &&
+						state.IsCompTaskStarted(id) &&
+						state.IsCompTaskCompleted(id)) {
+						validTransition = false
+					}
+				} else {
+					if !(state.IsTaskStarted(id) &&
+						state.IsTaskCompleted(id)) {
+						validTransition = false
+					}
+				}
+			}
+
+			// either we made a valid transition and had a valid update or applying
+			// this message is an invalidTransition and an error was returned.
+			validUpdate := validTransition && err == nil &&
+				newState != nil && newState.IsSagaCompleted()
+			errorReturned := !validTransition && err != nil && newState == nil
+
+			return validUpdate || errorReturned
+		},
+		GenSagaState(),
+	))
+
+	// Abort messages are valid unless a Saga has been Completed
+	properties.Property("AbortSaga message is valid or returns an error", prop.ForAll(
+		func(state *SagaState) bool {
+			validTransition := !state.IsSagaCompleted()
+
+			msg := MakeAbortSagaMessage(state.SagaId())
+			newState, err := updateSagaState(state, msg)
+
+			// either we made a valid transition and had a valid update or applying
+			// this message is an invalidTransition and an error was returned.
+			validUpdate := validTransition && err == nil &&
+				newState != nil && newState.IsSagaAborted()
+			errorReturned := !validTransition && err != nil && newState == nil
+
+			return validUpdate || errorReturned
+		},
+		GenSagaState(),
+	))
+
+	// StartTask messages are valid unless a Saga has been Completed or Aborted
```

Comment:
In what situation would we want to allow starting a task after it has already been ended? And starting a comp task after it has also ended?